util/privatepw/privatepw.8
.\"
.\" Copyright (c) 1999-2003 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"     $Id: privatepw.8,v 1.8 2009/04/19 10:35:43 wmaton Exp $
.\"
.TH PRIVATEPW 8
.SH NAME
privatepw \- Change WU\-FTPD Group Access File Information
.SH SYNOPSIS
.B privatepw
[-c] [-f ftpgroups] [-g group] accessgroup
.P
.B privatepw
[-d] [-f ftpgroups] accessgroup
.P
.B privatepw
[-l] [-f ftpgroups] 
.P
.B privatepw
[-V] 

.SH DESCRIPTION
If a site supports the SITE GROUP and SITE GPASS commands it is necessary
to be able to add, delete and list enhanced access group information. The
.I privatepw
utility is used to update this information in the group access 
(aka, ftpgroups) file. This is meant as an administrative tool
and not to be run by the general user population. This command 
requires read/write permission to the appropriate ftpgroups file.

.SH BACKGROUND

After user logs in, the SITE GROUP and SITE GPASS commands
may be used to specify an enhanced access group and associated
password. If the group name and password are valid, the user
becomes (via setegid()) a member of the group list in the group
access file (specified by the _PATH_PRIVATE define in pathnames.h).
.P
The format of the group access file is: 
.P
access_group_name:encrypted_password:real_group_name 
.P
where access_group_name is an arbitrary (alphanumeric +
punctuation) string. encrypted_password is the password encrypted
via crypt(3), exactly like in /etc/passwd. real_group_name is the
name of a valid group listed in /etc/group. 

.SH OPTIONS
.TP
.I \-c
Create a new ftpgroups file.
.TP
.I \-d
Delete the specified accessgroup's information from the ftpgroups file.
.TP
.I \-g group
Set the real system group to the group specified.  This is valid
only when adding a new accessgroup or changing an existing accessgroup's
password.
.TP
.I \-f ftpgroups
Use the specified ftpgroups file for all updates.
.TP
.I \-l
List the contents of the appropriate ftpgroups file.
.TP
.I \-?
Display a simple help usage message.
.TP
The
.I \-V
option causes the program to display copyright and version information, then
terminate.
.SH "SEE ALSO"
ftpaccess(5)
.SH BUGS
None known.


CHANGES
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 Stan Barber.
  Portions Copyright (c) 1997 Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: CHANGES,v 1.47 2016/03/14 23:57:54 wmaton Exp $

###########################################################################

Changes in 2.8.0-CC5: Released

 o  Updates to config.guess and config.sub to modernise the autoconf
    build system.

 o  configure script regenerated.

 o  Describe auto-buf tuning configure option in README.AUTOCONF.

 o  Add option (-R) to disable REST command in cases where download
    managers can abuse the server by opening multiple connections to
    retrieve chunks of the same file.  Based on a patch by Philippe Troin
    (See Debian bug ID 138497).

 o  Some minor man page cleanups.

 o  More syncs to Debian patches dating back to 2.4.2.

 o  Update INSTALL instructions.

 o  Changed the TLS default cipher from ALL:!EXP to exclude most insecure
    or unsupported ciphers within modern versions of OpenSSL.

Changes in 2.8.0-CC4: Released 7 January, 2012

 o  Thanks to a patch by Laurent CONSTANTIN, calling SSL_library_init()
    will no longer create an error within the SSL code related to a lack
    of ciphers.  The patch is applied to src/tlsutil.c.

 o  Another patch by Laurent CONSTANTIN, this time eliminating extra
    carriage returns that aren't necessary when calling syslog().

 o  Added -c option which displays the compile-time options that were used
    to build WU-FTPD. 

Changes in 2.8.0-CC3: Released 6 November, 2011

 o  Many URL updates.

 o  The default email address used by the server in absence of an email
    directive in ftpaccess is ftp-bugs, and is noted in ftpaccess now.

 o  Added code for ADLER-32 checksums for future extension to the list
    of algorithms available to do checksums.

 o  hard-coded links to cksum and md5sum within the FTP root now point to
    these utilities in /bin instead of /usr/bin.

 o  As defined in GFD.20 (GridFTP:  Protocol Extension to FTP for the
    Grid) by the Global Grid Forum, the SBUF command has been introduced.
    This calls the same code as SITE BUFSIZE would.  SITE BUFSIZE is
    still present for backwards compatibility.

 o  Thanks to a contribution by Austin Ellis, added RHEL6 packaging info
    to the codebase in packages/rpm.  This also includes a patch for
    building on that particular system.

 o  Updated ftpd.8 manpage documenting some missing SITE commands that
    had been in the server since 1998.

 o  First cut at implementing the remainder of the RFC3659 extended FTP
    commands.  FEAT, MLST, OPTS and MLSD are now supported.  Based upon
    early gridFTP releases which used WU-FTPD 2.6.0 and 2.6.1 code.

 o  According to draft-somers-ftp-mfxx-04.txt, defined the stubs for the
    MFF, MFCT and MFMT commands.  These may or may not be supported in the
    future, but at the least they are defined for now.  If so, they may
    become SITE commands instead if the spec is not approved.

 o  Updated doc/examples/ftpconversions based on Debian changes introduced
    sometime in 1995.  The server can now deal with such things as bzip2,
    xz, zip and unzip.

 o  If there's an empty /etc/fstab, a segfault can be produced.  A fix
    from Debian BugID 568235 to path_to_device in src/extensions.c addresses
    this.  Patch by Chris Butler.

 o  Added another possible path for quota.h, patch by Chris Butler of
    Debian.

 o  Added missing handler for SIGTERM, patch by Chris Butler of Debian
    related to Debian bug #536774.

 o  Changed -h invocation on the commandline to print out a very brief help
    listing all commandline options that could be supplied to the daemon.
    Also note that some switches are or have an effect on the configuration
    file.  The former function assigned to the 'h' option is now 'H'.

 o  The former function assigned to the 'h' option is now 'H'.  Said it
    twice now.  The manpage has been updated to reflect the changes.

 o  Implemented the CSID command as found in the IETF draft
    draft-peterson-streamlined-ftp-command-extensions.  This is meant
    to be a standards-based replacement for the CLNT command.  Currently
    this command is hard-wired with the server name and version, but the 
    next release will permit customisation.

###########################################################################

Changes in 2.8.0-CC2: Released 26 June, 2009

 o  Cleaned-up and updated the ftpd.8 manpage which had languished for
    some time.

 o  Added Auto tuning/auto buf, with added commands SITE BUFSIZE and
    SITE BUFSIZEMEASURE.  Use --enable-autobuf to enable this feature.
    Based on a patch on work done on WU-FTPD-2.6.1 by Gaurav Navlakha, et al
    of DAST, NLANR (now CAIDA).  This code also contains the ability to
    report back on buffer usage.  Note that an autobuf client needs to
    be used to take advantage of this feature.  For the time being, the
    code for buffer measurement is IPv4 only.

 o  Updated src/extensions.c to take into account a change in Linux quota
    structure.  Based on a patch by Chris Butler.


util/privatepw/privatepw.c
/**************************************************************************** 

   Copyright (c) 1999-2003 WU-FTPD Development Group. 
   All rights reserved.
   
   Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
   The Regents of the University of California.  Portions Copyright (c) 
   1993, 1994 Washington University in Saint Louis.  Portions Copyright 
   (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
   1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
   P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
   Technology.  Portions Copyright (c) 1997 by Stan Barber.  Portions 
   Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
   Foundation, Inc.  Portions Copyright (c) 1997 by Kent Landfield. 
 
   Use and distribution of this software and its source code are governed 
   by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
 
   $Id: privatepw.c,v 1.8 2009/04/19 10:35:43 wmaton Exp $
 
****************************************************************************/
/*
   Subsystem:  WU-FTPD FTP Server
   Purpose:    Change WU-FTPD Guest Passwords
   File Name:  privatepw.c               

   usage: privatepw [-c] [-f passwordfile] [-g group] accessgroup
   privatepw [-d] [-f passwordfile] accessgroup
   privatepw [-l] [-f passwordfile] 
   -c:           creates a new file.
   -d:           deletes specified accessgroup.
   -l:           list contents of ftpgroups file.
   -f ftpgroups: updates the specified file.
   -g group:     set real group to the specified group.

   This software was initially written by Kent Landfield (kent@landfield.com)
 */

#include "../../src/config.h"
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/signal.h>
#include <sys/stat.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <grp.h>
#include <unistd.h>
#include "../../src/pathnames.h"

#define BUFLEN 256
#define GROUPLEN 8

static char *tmp;
static char line[BUFLEN];
static FILE *fp;
static int verbose = 0;
static char *passwdpath;

static unsigned char itoa64[] =	/* 0 ... 63 => ascii - 64 */
"./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

void print_copyright(void);

static void usage(void)
{
    fprintf(stderr, "usage: privatepw [-c] [-f ftpgroups] [-g group] accessgroup\n");
    fprintf(stderr, "       privatepw [-d] [-f ftpgroups] accessgroup\n");
    fprintf(stderr, "       privatepw [-l] [-f ftpgroups]\n");
    fprintf(stderr, "\t\t-c:           creates a new file.\n");
    fprintf(stderr, "\t\t-d:           deletes specified accessgroup.\n");
    fprintf(stderr, "\t\t-l:           list contents of ftpgroups file.\n");
    fprintf(stderr, "\t\t-f ftpgroups: updates the specified file.\n");
    fprintf(stderr, "\t\t-g group:     set real group to the specified group.\n");
    exit(EXIT_SUCCESS);
}

static void to64(register char *s, register long v, register int n)
{
    while (--n >= 0) {
	*s++ = itoa64[v & 0x3f];
	v >>= 6;
    }
}

static void terminate(void)
{
    if (tmp)
	(void) unlink(tmp);
    exit(EXIT_FAILURE);
}

static void catchintr(void)
{
    fprintf(stderr, "Interrupted.\n");
    terminate();
}

static char *savit(char *s)
{
    char *d;
    int len;

    if (s == NULL)
	terminate();

    len = strlen(s)+1;

    if ((d = (char *) malloc(len)) == NULL) {
	fprintf(stderr, "Whoa... Malloc failed.\n");
	terminate();
    }
    strlcpy(d, s, len);
    return (d);
}

static int confirmed(char *accessgroup)
{
    register int ch;

    printf("Delete %s: Are your sure ? (y/n) ", accessgroup);
    ch = getc(stdin);
    if (ch == 'y')
	return (1);
    return (0);
}

static char *getgroup(char *msg)
{
    register int ch;
    register char *p;
    static char buf[GROUPLEN + 1];

    (void) fputs(msg, stderr);
    rewind(stderr);		/* implied flush */
    for (p = buf; (ch = getc(stdin)) != EOF && ch != '\n';)
	if (p < buf + GROUPLEN)
	    *p++ = ch;

    if (p > buf+GROUPLEN) {
	fprintf(stderr, "group name too long...\n");
	terminate();
    }
    else
        *p = '\0';

    if (getgrnam(buf) == NULL) {
	fprintf(stderr, "Invalid group \'%s\' specified\n", buf);
	terminate();
    }
    return (buf);
}

#if !defined(NO_CRYPT_PROTO)
extern char *crypt(const char *, const char *);
#endif /* !defined(NO_CRYPT_PROTO) */ 
extern char *getpass(const char *prompt);

static void addrecord(char *accessgroup, char *sysgroup, size_t slen, char *msg, FILE *f)
{
    char *pw, *cpw, salt[3];

    printf("%s %s\n", msg, accessgroup);

    if (sysgroup[0] == '\0')
	strlcpy(sysgroup, getgroup("Real System Group to use: "), slen);

    pw = savit((char *) getpass("New password: "));
    if (strcmp(pw, (char *) getpass("Re-type new password: "))) {
	fprintf(stderr, "They don't match, sorry.\n");
	if (tmp)
	    unlink(tmp);
        free(pw);
	exit(EXIT_FAILURE);
    }

    srand((int) time((time_t *) NULL));
    to64(&salt[0], rand(), 2);
    cpw = crypt(pw, salt);
    free(pw);
    fprintf(f, "%s:%s:%s\n", accessgroup, cpw, sysgroup);
}

static void list_privatefile(char *privatefile)
{
    if (verbose)
	fprintf(stderr, "Private File: %s file.\n", privatefile);

    if ((fp = fopen(privatefile, "r")) == NULL) {
	fprintf(stderr, "Could not open %s file.\n", privatefile);
	exit(EXIT_FAILURE);
    }

    printf("\nWU-FTPD Private file: %s\n", privatefile);
    printf("accessgroup : password : system group\n");
    printf("-------\n");

    while (fgets(line, BUFLEN, fp) != NULL)
	fputs(line, stdout);
    printf("-------\n");
}

int main(int argc, char **argv)
{
    extern void (*signal(int sig, void (*disp) (int))) (int);
    extern int getopt(int argc, char *const *argv, const char *optstring);
    extern char *optarg;
    extern int optind;
    extern int opterr;

    struct stat stbuf;

    char realgroup[BUFLEN];
    char *cp;

    char accessgroup[BUFLEN];
    char w[BUFLEN];
    char *command[4];

    int create;
    int delete;
    int list;
    int found;
    int lineno;
    int c;

    FILE *tfp;
    int tfd;
    char tmpname[BUFLEN];

    pid_t cp_pid;

    opterr = 0;
    create = 0;
    delete = 0;
    list = 0;

    tmp = NULL;
    realgroup[0] = '\0';

    passwdpath = NULL;

    if (argc == 1)
	usage();

    while ((c = getopt(argc, argv, "Vvcdf:g:l")) != EOF) {
	switch (c) {
	case 'd':
	    delete++;
	    break;
	case 'c':
	    create++;
	    break;
	case 'f':
	    passwdpath = strdup(optarg);
	    break;
	case 'g':
	    /* Let's make sure our group name isn't longer than BUFLEN */
	    if (strlen(optarg) > BUFLEN) {
		fprintf(stderr, "Group name too long.\n");
		return(1);
	    } else
		strlcpy(realgroup, optarg, sizeof(realgroup));
	    if (getgrnam(realgroup) == NULL) {
		fprintf(stderr, "Invalid group \'%s\' specified\n", realgroup);
		return (1);
	    }
	    break;
	case 'l':
	    list++;
	    break;
	case 'v':
	    verbose++;
	    break;
	case 'V':
	    print_copyright();
	    return (0);
	    /* NOTREACHED */
	default:
	    usage();
	}
    }

    if (passwdpath == NULL)  /* set the default path */
       passwdpath = strdup(_PATH_PRIVATE);

    if (list) {
	list_privatefile(passwdpath);
	return (0);
    }

    if (optind >= argc) {
	fprintf(stderr, "Need to specify an accessgroup name.\n");
	usage();
    }

    signal(SIGINT, (void (*)()) catchintr);

    /* Let's check to make sure argv[optind] isn't larger than BUFLEN
     * and if it is there is no point in continuing since simply
     * truncating the group name is useless, because it results in
     * a group name other than that intended by the user
     */
    if(strlen(argv[optind]) > BUFLEN) {
	fprintf(stderr, "Specified accessgroup name is too large.\n");
	return(1);
    } else
	strlcpy(accessgroup, argv[optind], sizeof(accessgroup));

    if (create) {
	if (stat(passwdpath, &stbuf) == 0) {
	    fprintf(stderr, "%s exists, cannot create it.\n", passwdpath);
	    fprintf(stderr, "Remove -c option or use the -f option to specify another.\n");
	    return (1);
	}

	if ((tfp = fopen(passwdpath, "w")) == NULL) {
	    fprintf(stderr, "Could not open \"%s\" for writing.\n", passwdpath);
	    perror("fopen");
	    return (1);
	}

	tmp = passwdpath;

	printf("Creating WU-FTPD Private file: %s\n", passwdpath);
	addrecord(accessgroup, realgroup, sizeof(realgroup), "Adding accessgroup", tfp);

	fclose(tfp);
	return (0);
    }

#if defined(HAVE_MKSTEMP)
    strlcpy(tmpname, "/tmp/privatepwXXXXXX", sizeof(tmpname));
    tmp = tmpname;
    tfd = mkstemp(tmp);
    if(tfd < 0) {
	fprintf(stderr, "Could not open temp file.\n");
	return(1);
    }
#else /* !(defined(HAVE_MKSTEMP)) */ 
#  if defined(HAVE_MKTEMP)
    tmp = mktemp("/tmp/privatepwXXXXXX");
#  else /* !(defined(HAVE_MKTEMP)) */ 
    tmp = tmpnam(NULL);
#  endif /* !(defined(HAVE_MKTEMP)) */ 
#endif /* !(defined(HAVE_MKSTEMP)) */ 

#if defined(HAVE_MKSTEMP)
    if ((tfp = fdopen(tfd, "w")) == NULL) {
        unlink(tmp);
#else /* !(defined(HAVE_MKSTEMP)) */ 
    if ((tfp = fopen(tmp, "w")) == NULL) {
#endif /* !(defined(HAVE_MKSTEMP)) */ 
	fprintf(stderr, "Could not open temp file.\n");
	return (1);
    }

    if ((fp = fopen(passwdpath, "r")) == NULL) {
	fprintf(stderr, "Could not open %s file.\n", passwdpath);
	fprintf(stderr, "Use -c option to create new one.\n");
	return (1);
    }

    lineno = 0;
    found = 0;

    while (fgets(line, BUFLEN, fp) != NULL) {
	lineno++;

	if (found || (line[0] == '#') || (!line[0])) {
	    fputs(line, tfp);
	    continue;
	}

	strlcpy(w, line, sizeof(w));

	if ((cp = strchr(w, ':')) == NULL) {
	    fprintf(stderr, "%s: line %d: invalid record format.\n", passwdpath, lineno);
	    continue;
	}
	*cp++ = '\0';

	if ((cp = strchr(cp, ':')) == NULL) {
	    fprintf(stderr, "%s: line %d: invalid record format.\n", passwdpath, lineno);
	    continue;
	}
	*cp++ = '\0';

	if (strcmp(accessgroup, w)) {
	    fputs(line, tfp);
	    continue;
	}
	else {
	    if (delete) {
		if (!confirmed(accessgroup))
		    terminate();
	    }
	    else {
		if (realgroup[0] == '\0') {
		    strlcpy(realgroup, cp, BUFLEN);
		    if ((cp = strchr(realgroup, '\n')) != NULL)
			*cp = '\0';
		}
		addrecord(accessgroup, realgroup, sizeof(realgroup), "Updating accessgroup", tfp);
	    }
	    found = 1;
	}
    }

    if (!found && !delete)
	addrecord(accessgroup, realgroup, sizeof(realgroup), "Adding accessgroup", tfp);
    else if (!found && delete) {
	fprintf(stderr, "%s not found in %s.\n", accessgroup, passwdpath);
	terminate();
    }

    fclose(fp);
    fclose(tfp);

    command[0] = "cp";
    command[1] = tmp;
    command[2] = passwdpath;
    command[3] = NULL;

    cp_pid = fork();
    if(cp_pid==0) /* child -> execvp() cp */
        execvp("/bin/cp", command);
    else {
	int status;
	waitpid(cp_pid, &status, 0);
    }
    (void) unlink(tmp);
    exit (EXIT_SUCCESS);
}
